zkevm-circuits/src/circuit_tools/memory.rs

```diff
@@ -1,47 +1,64 @@
 //! Memory
-use crate::util::{query_expression, Expr};
+use crate::{
+    evm_circuit::util::rlc,
+    util::{query_expression, Expr},
+};
 use eth_types::Field;
 use halo2_proofs::{
     circuit::Value,
-    plonk::{Advice, Column, ConstraintSystem, Error, Expression},
+    plonk::{
+        Advice, Column, ConstraintSystem, Error, Expression, FirstPhase, SecondPhase, ThirdPhase,
+    },
     poly::Rotation,
 };
-use std::ops::{Index, IndexMut};
+use itertools::Itertools;
+use std::{
+    collections::HashMap,
+    ops::{Index, IndexMut},
+};
 
 use super::{
-    cached_region::CachedRegion, cell_manager::CellType, constraint_builder::ConstraintBuilder,
+    cached_region::CachedRegion,
+    cell_manager::{Cell, CellType},
+    constraint_builder::ConstraintBuilder,
 };
 
 #[derive(Clone, Debug, Default)]
 pub(crate) struct Memory<F, C> {
-    columns: Vec<Column<Advice>>,
+    height: usize,
     banks: Vec<MemoryBank<F, C>>,
+    rw_records: HashMap<C, (Column<Advice>, Column<Advice>)>,
```

Comment:
Why not store this on the MemoryBank level?

Reply:
I guess it's because you had the `key: Column<Advice>` on the memory level before haha, but there's usage of it when building lookups cuz I can iterate through the records within the Memory without going in. But either way works, I can change it if necessary.

Reply:
Ah yeah the idea was that everything would create its own columns automatically but that's of course not yet how that worked. :)

I think it would be cool if the MemoryBank does it because I saw that a MemoryBank being just one possible implementation to implement a memory system, you can think of other ones (like a stack based one or whatever) that could be used instead. By moving the logic into the MemoryBank we can easily have different types of memory.

Reply:
Cool idea. I'll move the lookup back with the `add_lookup` and only leave abstraction in the memory bank level, and then I think for the next PR related to this we can do different creative impl!